packages/x-date-pickers/src/hooks/useClearableField.tsx

```diff
@@ -0,0 +1,113 @@
+  FieldSlotsComponentsProps,
+  FieldsTextFieldProps,
+  useLocaleText,
+} from '../internals';
+
+type UseClearableFieldProps<
+  TFieldProps extends FieldsTextFieldProps,
+  TInputProps extends { endAdornment?: React.ReactNode } | undefined,
+  TFieldSlots extends FieldSlotsComponents,
+  TFieldSlotsComponentsProps extends FieldSlotsComponentsProps,
+> = {
+  clearable: boolean;
+  fieldProps: TFieldProps;
+  InputProps: TInputProps;
+  onClear: React.MouseEventHandler<HTMLButtonElement>;
+  slots?: { [K in keyof TFieldSlots as Uncapitalize<K & string>]: TFieldSlots[K] };
+  slotProps?: TFieldSlotsComponentsProps;
+  components?: TFieldSlots;
+  componentsProps?: TFieldSlotsComponentsProps;
+};
+
+export const useClearableField = <
+  TFieldProps extends FieldsTextFieldProps,
+  TInputProps extends { endAdornment?: React.ReactNode } | undefined,
+  TFieldSlotsComponents extends FieldSlotsComponents,
+  TFieldSlotsComponentsProps extends FieldSlotsComponentsProps,
+>({
+  clearable,
+  fieldProps: forwardedFieldProps,
+  InputProps: ForwardedInputProps,
+  onClear,
+  slots,
+  slotProps,
+  components,
+  componentsProps,
+}: UseClearableFieldProps<
+  TFieldProps,
+  TInputProps,
+  TFieldSlotsComponents,
+  TFieldSlotsComponentsProps
+>) => {
+  const localeText = useLocaleText();
+
+  const IconButton = slots?.clearButton ?? components?.ClearButton ?? MuiIconButton;
+  // The spread is here to avoid this bug mui/material-ui#34056
+  const { ownerState, ...iconButtonProps } = useSlotProps({
+    elementType: IconButton,
+    externalSlotProps: slotProps?.clearButton ?? componentsProps?.clearButton,
+    ownerState: {},
+    className: 'clearButton',
+    additionalProps: {
+      title: localeText.fieldClearLabel,
+    },
+  });
+  const EndClearIcon = slots?.clearIcon ?? components?.ClearIcon ?? ClearIcon;
+  const endClearIconProps = useSlotProps({
+    elementType: EndClearIcon,
+    externalSlotProps: slotProps?.clearIcon ?? componentsProps?.clearIcon,
+    ownerState: {},
+  });
+
+  const InputProps = {
+    ...ForwardedInputProps,
+    endAdornment: clearable ? (
+      <React.Fragment>
+        <InputAdornment
+          position="end"
+          sx={{ marginRight: ForwardedInputProps?.endAdornment ? -1 : -1.5 }}
+        >
+          <IconButton {...iconButtonProps} onClick={onClear}>
+            <EndClearIcon fontSize="small" {...endClearIconProps} />
+          </IconButton>
+        </InputAdornment>
+        {ForwardedInputProps?.endAdornment}
+      </React.Fragment>
+    ) : (
+      ForwardedInputProps?.endAdornment
+    ),
+  };
+
+  const fieldProps: TFieldProps = {
+    ...forwardedFieldProps,
+    sx: [
+      {
+        '& .clearButton': {
+          visibility: 'visible',
+        },
+        '@media (pointer: fine)': {
+          '& .clearButton': {
+            visibility: 'hidden',
+          },
+          '&:hover, &:focus-within': {
+            '.clearButton': {
+              visibility: 'visible',
```

Comment:
One final suggestion.
WDYT about using `opacity` or any other way of only visually hiding the button to not interfere with screen reader behavior? 🤔 
`visibility: hidden` is treated as a hidden element and when doing <kbd>Shift</kbd>+<kbd>Tab</kbd> while using a screen reader on a filled Field element - the focus would jump to the input instead of the clear value button as it would doing the same in a regular environment. 🙈 
```suggestion
          opacity: 1
        },
        '@media (pointer: fine)': {
          '& .clearButton': {
            opacity: 0,
          },
          '&:hover, &:focus-within': {
            '.clearButton': {
              opacity: 1,
```